src/sprightly.ts

```diff
@@ -1,3 +1,60 @@
-export = function sprightly() {
-  console.log("sprightly")
+import { readFile } from "fs/promises"
+
+interface Data {
+  [key: string]: string | number | Array<string | number | Data> | Data
 }
+
+type Options = {
+  cache?: boolean
+  keyFallback?: string
+  throwOnKeyNotfound?: boolean
+}
+
+const defaultOptions: Options = {
+  keyFallback: "",
+  throwOnKeyNotfound: false,
+  cache: false,
+}
+
+export = async function sprightly(
+  entryPoint: string,
+  data: Data,
+  options: Options = defaultOptions,
+): Promise<string> {
+  const regex = /<<(.*?)>>|\{\{(.*?)\}\}/gi
+  const file = (await readFile(entryPoint)).toString()
+  const result = file.replaceAll(regex, (match) => {
```

Comment:
So ... the way you handled parsing the file is pretty neat. Though here's a suggestion for an implementation that might seem better:

As the callback for the `.replaceAll()` function, you should have this (this callback to `.replaceAll()` assumes that the regex is the one suggested above, you might have to modify it if you decided to go with a different one):
```js
(_, isComponent: string, reference: string) => {
  if (isComponent) {
    // logic to parse component
    return
  }

  /** input-output samples that explain how the function works
  * "key" => ["key"]
  * "property.nested[0].anotherProperty[1]" => ["property", "nested", 0, "anotherProperty", 1]
  */
  const path = subscriptingToPath(reference)

  // rest of logic to read the property using the path. I would suggest using `Array.reduce`, it's perfect for this situation.
}
```

A note about `subscriptingToPath` implementation:
Investigate [this regex](https://regex101.com/r/f4Y8zW/1) that could reduce the whole implementation of this function into 1 line, toggle between the different flavors of C# and JS, observe the Match Information widget, and try not to hate JS 😃.

Reply:
https://regex101.com/r/KjfuU4/1